src/receiver.rs

```diff
@@ -6,68 +6,79 @@ use core::task::{Context, Poll};
 /// The receiving half of a oneshot channel.
 #[derive(Debug)]
 pub struct Receiver<T> {
-    inner: Arc<Inner<T>>,
-    done: bool,
+    x: Option<Arc<Inner<T>>>,
 }
 
 impl<T> Receiver<T> {
-
     pub(crate) fn new(inner: Arc<Inner<T>>) -> Self {
-        Receiver { inner, done: false }
+        Receiver { x: Some(inner) }
     }
 
     /// Closes the channel by causing an immediate drop.
-    pub fn close(self) { }
+    pub fn close(self) {}
 
     /// Attempts to receive. On failure, if the channel is not closed,
     /// returns self to try again.
     pub fn try_recv(mut self) -> Result<T, TryRecvError<T>> {
-        let state = self.inner.state();
+        if let Some(x) = self.x.take() {
+            let state = x.state();
+            if state.ready() {
+                Ok(x.take_value())
+            } else if state.closed() {
+                Err(TryRecvError::Closed)
+            } else {
+                Err(TryRecvError::Empty(Self { x: Some(x) }))
+            }
+        } else {
+            Err(TryRecvError::Closed)
+        }
+    }
+
+    fn handle_state(inner: &Inner<T>, state: crate::inner::State) -> Poll<Result<T, Closed>> {
         if state.ready() {
-            self.done = true;
-            Ok(self.inner.take_value())
+            Poll::Ready(Ok(inner.take_value()))
         } else if state.closed() {
-            self.done = true;
-            Err(TryRecvError::Closed)
+            Poll::Ready(Err(Closed()))
         } else {
-            Err(TryRecvError::Empty(self))
+            Poll::Pending
         }
     }
 }
 
-
-
 impl<T> Future for Receiver<T> {
     type Output = Result<T, Closed>;
     fn poll(self: Pin<&mut Self>, ctx: &mut Context) -> Poll<Result<T, Closed>> {
         let this = Pin::into_inner(self);
-        let state = this.inner.state();
-        if state.ready() {
-            this.done = true;
-            Poll::Ready(Ok(this.inner.take_value()))
-        } else if state.closed() {
-            this.done = true;
-            Poll::Ready(Err(Closed()))
-        } else {
-            let state = this.inner.set_recv(ctx.waker().clone());
-            if state.ready() {
-                this.done = true;
-                Poll::Ready(Ok(this.inner.take_value()))
-            } else {
-                if state.send() { this.inner.send().wake_by_ref(); }
-                Poll::Pending
```

Comment:
P.S. I have a question about this part: Isn't here a race condition that the channel could be closed after the `this.inner.state()` call, but before the `this.inner.set_recv` call, which would mean that the Receiver would never wake up... (because it doesn't check `state.closed()` again)

Reply:
yup, exactly right